File: surface/widgets/vue/components/widgets/concepts/llm-safety/ToolCallDetail.ts
import { defineComponent, h, ref, computed, type PropType } from 'vue';

export type ToolCallDetailState = 'idle' | 'retrying';
export type ToolCallDetailEvent =
  | { type: 'EXPAND_ARGS' }
  | { type: 'EXPAND_RESULT' }
  | { type: 'RETRY' }
  | { type: 'RETRY_COMPLETE' }
  | { type: 'RETRY_ERROR' };

export function toolCallDetailReducer(state: ToolCallDetailState, event: ToolCallDetailEvent): ToolCallDetailState {
  switch (state) {
    case 'idle':
      if (event.type === 'EXPAND_ARGS') return 'idle';
      if (event.type === 'EXPAND_RESULT') return 'idle';
      if (event.type === 'RETRY') return 'retrying';
      return state;
    case 'retrying':
      if (event.type === 'RETRY_COMPLETE') return 'idle';
      if (event.type === 'RETRY_ERROR') return 'idle';
      return state;
    default:
      return state;
  }
}

function formatJson(str: string): string {
  try { return JSON.stringify(JSON.parse(str), null, 2); } catch { return str; }
}

export const ToolCallDetail = defineComponent({
  name: 'ToolCallDetail',
  props: {
    toolName: { type: String, required: true },
    arguments: { type: String, required: true },
    result: { type: String, default: undefined },
    timing: { type: Number, default: undefined },
    tokenUsage: { type: Number, default: undefined },
    error: { type: String, default: undefined },
    showTiming: { type: Boolean, default: true },
    showTokens: { type: Boolean, default: true },
  },
  emits: ['retry'],
  setup(props, { emit }) {
    const state = ref<ToolCallDetailState>('idle');
    const argsExpanded = ref(true);
    const resultExpanded = ref(true);

    function send(event: ToolCallDetailEvent) {
      state.value = toolCallDetailReducer(state.value, event);
    }

    const hasError = computed(() => !!props.error);
    const statusLabel = computed(() => {
      if (state.value === 'retrying') return 'Retrying...';
      if (hasError.value) return 'Error';
      if (props.result != null) return 'Success';
      return 'Pending';
    });

    function handleRetry() {
      send({ type: 'RETRY' });
      emit('retry');
    }

    function handleKeydown(e: KeyboardEvent) {
      if (e.key === 'r' && e.ctrlKey) { e.preventDefault(); handleRetry(); }
    }

    return () => {
      const children: any[] = [];

      // Header
      children.push(h('div', { 'data-part': 'header' }, [
        h('span', { 'data-part': 'tool-name' }, props.toolName),
        h('div', { 'data-part': 'status-badge', 'data-status': hasError.value ? 'error' : 'success' }, statusLabel.value),
      ]));

      // Arguments panel
      children.push(h('div', { 'data-part': 'arguments-panel' }, [
        h('button', {
          type: 'button', 'data-part': 'section-toggle',
          'aria-expanded': argsExpanded.value ? 'true' : 'false',
          onClick: () => { argsExpanded.value = !argsExpanded.value; send({ type: 'EXPAND_ARGS' }); },
        }, `${argsExpanded.value ? '\u25BC' : '\u25B6'} Arguments`),
        argsExpanded.value ? h('pre', { 'data-part': 'arguments-json' }, formatJson(props.arguments)) : null,
      ]));

      // Result panel
      if (props.result != null) {
        children.push(h('div', { 'data-part': 'result-panel' }, [
          h('button', {
            type: 'button', 'data-part': 'section-toggle',
            'aria-expanded': resultExpanded.value ? 'true' : 'false',
            onClick: () => { resultExpanded.value = !resultExpanded.value; send({ type: 'EXPAND_RESULT' }); },
          }, `${resultExpanded.value ? '\u25BC' : '\u25B6'} Result`),
          resultExpanded.value ? h('pre', { 'data-part': 'result-json' }, formatJson(props.result)) : null,
        ]));
      }

      // Timing
      if (props.showTiming && props.timing != null) {
        children.push(h('div', { 'data-part': 'timing-bar' }, [
          h('span', { 'data-part': 'timing-label' }, 'Duration'),
          h('span', { 'data-part': 'timing-value' }, `${props.timing}ms`),
        ]));
      }

      // Token badge
      if (props.showTokens && props.tokenUsage != null) {
        children.push(h('div', { 'data-part': 'token-badge' }, `${props.tokenUsage} tokens`));
      }

      // Error panel
      if (hasError.value) {
        children.push(h('div', { 'data-part': 'error-panel', role: 'alert' }, props.error));
      }

      // Retry button
      if (hasError.value) {
        children.push(h('button', {
          type: 'button', 'data-part': 'retry-button',
          disabled: state.value === 'retrying',
          'aria-label': 'Retry tool call',
          onClick: handleRetry,
        }, state.value === 'retrying' ? 'Retrying...' : 'Retry'));
      }

      return h('div', {
        role: 'article',
        'aria-label': `Tool call: ${props.toolName}`,
        'data-surface-widget': '',
        'data-widget-name': 'tool-call-detail',
        'data-part': 'root',
        'data-state': state.value,
        tabindex: 0,
        onKeydown: handleKeydown,
      }, children);
    };
  },
});

export default ToolCallDetail;

File: surface/widgets/swiftui/components/widgets/concepts/llm-safety/ToolCallDetail.swift
import SwiftUI

enum ToolCallDetailWidgetState {
    case idle, retrying
}

enum ToolCallDetailEvent {
    case expandArgs, expandResult, retry, retryComplete, retryError
}

func toolCallDetailReduce(state: ToolCallDetailWidgetState, event: ToolCallDetailEvent) -> ToolCallDetailWidgetState {
    switch state {
    case .idle:
        if event == .retry { return .retrying }
        return state
    case .retrying:
        if event == .retryComplete { return .idle }
        if event == .retryError { return .idle }
        return state
    }
}

struct ToolCallDetailView: View {
    var toolName: String
    var input: String
    var output: String? = nil
    var status: String = "pending"
    var duration: Int? = nil
    var timestamp: String? = nil
    var arguments: String = ""
    var result: String? = nil
    var timing: Int? = nil
    var tokenUsage: Int? = nil
    var error: String? = nil
    var showTiming: Bool = true
    var showTokens: Bool = true
    var onRetry: (() -> Void)? = nil

    @State private var widgetState: ToolCallDetailWidgetState = .idle
    @State private var argsExpanded: Bool = true
    @State private var resultExpanded: Bool = true

    private var resolvedInput: String { input.isEmpty ? arguments : input }
    private var resolvedOutput: String? { output ?? result }
    private var resolvedDuration: Int? { duration ?? timing }
    private var resolvedStatus: String { error != nil ? "error" : status }
    private var errorMessage: String? { error ?? (resolvedStatus == "error" && resolvedOutput != nil ? resolvedOutput : nil) }

    private func formatJson(_ value: String) -> String {
        guard let data = value.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data),
              let pretty = try? JSONSerialization.data(withJSONObject: obj, options: .prettyPrinted),
              let str = String(data: pretty, encoding: .utf8)
        else { return value }
        return str
    }

    private var statusColor: Color {
        switch resolvedStatus {
        case "success": return .green
        case "error": return .red
        case "pending": return .orange
        default: return .secondary
        }
    }

    private var statusLabel: String {
        switch resolvedStatus {
        case "success": return "Success"
        case "error": return "Error"
        case "pending": return "Pending"
        default: return resolvedStatus.capitalized
        }
    }

    private var statusBgColor: Color {
        switch resolvedStatus {
        case "success": return Color.green.opacity(0.15)
        case "error": return Color.red.opacity(0.15)
        case "pending": return Color.orange.opacity(0.15)
        default: return Color.secondary.opacity(0.1)
        }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Header
            HStack(spacing: 8) {
                Text(toolName)
                    .font(.system(.body, design: .monospaced))
                    .fontWeight(.semibold)

                Text(statusLabel)
                    .font(.caption)
                    .fontWeight(.medium)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 2)
                    .background(statusBgColor)
                    .foregroundColor(statusColor)
                    .cornerRadius(12)
                    .accessibilityLabel("Status: \(statusLabel)")

                Spacer()

                if let dur = resolvedDuration, showTiming {
                    Text("\(dur)ms")
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 10)

            Divider()

            // Input section (collapsible)
            VStack(alignment: .leading, spacing: 0) {
                Button {
                    withAnimation { argsExpanded.toggle() }
                } label: {
                    HStack(spacing: 6) {
                        Image(systemName: argsExpanded ? "chevron.down" : "chevron.right")
                            .font(.caption)
                        Text("Input")
                            .font(.subheadline)
                            .fontWeight(.semibold)
                        Spacer()
                    }
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .accessibilityLabel(argsExpanded ? "Collapse input" : "Expand input")

                if argsExpanded {
                    Text(formatJson(resolvedInput))
                        .font(.system(.caption, design: .monospaced))
                        .padding(.horizontal, 12)
                        .padding(.bottom, 8)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(Color.gray.opacity(0.05))
                        .accessibilityLabel("Arguments")
                }
            }

            Divider()

            // Output section (collapsible)
            if resolvedOutput != nil || errorMessage != nil {
                VStack(alignment: .leading, spacing: 0) {
                    Button {
                        withAnimation { resultExpanded.toggle() }
                    } label: {
                        HStack(spacing: 6) {
                            Image(systemName: resultExpanded ? "chevron.down" : "chevron.right")
                                .font(.caption)
                            Text("Output")
                                .font(.subheadline)
                                .fontWeight(.semibold)
                            Spacer()
                        }
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 8)
                    .accessibilityLabel(resultExpanded ? "Collapse output" : "Expand output")

                    if resultExpanded {
                        if resolvedStatus == "error", let errMsg = errorMessage {
                            Text(errMsg)
                                .font(.system(.caption, design: .monospaced))
                                .foregroundColor(.red)
                                .padding(.horizontal, 12)
                                .padding(.bottom, 8)
                                .frame(maxWidth: .infinity, alignment: .leading)
                                .background(Color.red.opacity(0.05))
                                .overlay(
                                    Rectangle().fill(Color.red).frame(width: 3),
                                    alignment: .leading
                                )
                                .accessibilityRole(.alert)
                                .accessibilityLabel("Error details")
                        } else if let out = resolvedOutput {
                            Text(formatJson(out))
                                .font(.system(.caption, design: .monospaced))
                                .padding(.horizontal, 12)
                                .padding(.bottom, 8)
                                .frame(maxWidth: .infinity, alignment: .leading)
                                .background(Color.gray.opacity(0.05))
                                .accessibilityLabel("Result")
                        }
                    }
                }

                Divider()
            }

            // Token usage badge
            if showTokens, let usage = tokenUsage {
                Text("\(usage) tokens")
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 2)
                    .overlay(
                        Capsule().stroke(Color.gray.opacity(0.3), lineWidth: 1)
                    )
                    .padding(.horizontal, 12)
                    .padding(.top, 8)
            }

            // Timestamp
            if let ts = timestamp {
                Text(ts)
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .padding(.horizontal, 12)
                    .padding(.top, 4)
                    .padding(.bottom, 8)
            }

            // Retry button
            if errorMessage != nil {
                Button(widgetState == .retrying ? "Retrying..." : "Retry") {
                    guard widgetState != .retrying else { return }
                    widgetState = toolCallDetailReduce(state: widgetState, event: .retry)
                    onRetry?()
                }
                .disabled(widgetState == .retrying)
                .padding(.horizontal, 12)
                .padding(.bottom, 12)
                .accessibilityLabel("Retry tool call")
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Tool call: \(toolName)")
    }
}
